### DISCUSSION TITLE
Run machines in another thread

### DISCUSSION THREAD

**Discussion by CodingDive** [statelyai/xstate#1255] [General]
Is it possible to move all the guards and assign actions off the main-thread?

Has anyone tried running machines in web workers with a simple "send" interface?

**Comment by Andarist** [comment]
> Is it possible to move all the guards and assign actions off the main-thread?

No. Or rather - impossible for guards unless you make all guarded transitions "multi-state". As you have noticed - everything would have to be communicated through events between your threads. So it's totally doable but would complicate your models quite significantly. I'm not sure what benefits of this would be. I wouldn't expect machines to be a bottleneck in your application 😉 

If you feel adventurous you could experiment with that but I wouldn't personally recommend trying to think in terms of guards and actions, just sending events back and forth. Later you could try to build some abstraction that would hide this complexity in it, but I would try to implement bare-bones communication based on events first to check if it brings any benefits to your use case.

**Reply by CodingDive** [reply]
I was thinking of running the whole machine in a service worker. I just mentioned guards and actions since those are the two statechart building blocks most likely to block the main thread. A lot of actions or guards in some code of mine are O(n) or O(n2) and if the same machine manages user input and a bunch of other things, it can quickly lead to a laggy user experience.
